File: src/k8s/kubectl.rs
use std::collections::HashMap;
use std::str::FromStr;

use gigurra_rust_util as util;
use util::shell::RunCommandError;

use crate::ApiResourceType;

#[derive(Debug, PartialEq, Clone)]
pub struct ApiResourceTypes {
    pub all: Vec<ApiResourceType>,
    pub accessible: AccessibleApiResourceTypes,
}

impl ApiResourceTypes {
    pub fn from(all_values: &[ApiResourceType]) -> Self {
        Self {
            all: Vec::from(all_values),
            accessible: AccessibleApiResourceTypes::from(all_values),
        }
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct AccessibleApiResourceTypes {
    pub all: Vec<ApiResourceType>,
    pub namespaced: Vec<ApiResourceType>,
    pub global: Vec<ApiResourceType>,
}

impl AccessibleApiResourceTypes {
    pub fn from(all_values: &[ApiResourceType]) -> Self {
        let accessible_resources =
            all_values
                .iter()
                .filter(|x| x.verbs.contains(&"get".to_string()))
                .cloned()
                .collect::<Vec<ApiResourceType>>();

        Self {
            all: accessible_resources.clone(),
            namespaced: accessible_resources.iter().filter(|x| x.namespaced).cloned().collect::<Vec<ApiResourceType>>(),
            global: accessible_resources.iter().filter(|x| !x.namespaced).cloned().collect::<Vec<ApiResourceType>>(),
        }
    }
}

pub fn api_resource_types() -> Result<ApiResourceTypes, RunCommandError> {
    let result = util::shell::run_command(
        std::process::Command::new("kubectl").arg("api-resources").arg("-o").arg("wide")
    );

    result.map(|command_result_output| {
        let lines: Vec<String> = command_result_output.lines().map(String::from).collect::<Vec<String>>();
        let line_maps: Vec<HashMap<String, String>> = util::string::parse_stdout_table(&lines);
        let list: Vec<ApiResourceType> = line_maps.iter().map(map_to_resource_type).collect::<Vec<ApiResourceType>>();

        ApiResourceTypes::from(&list)
    })
}

pub fn download_everything(types_to_download: &[&ApiResourceType]) -> Result<String, RunCommandError> {
    let qualified_names: Vec<String> = types_to_download.iter().map(|x| x.qualified_name()).collect();
    let qualified_names_joined: String = qualified_names.join(",");
    util::shell::run_command(
        std::process::Command::new("kubectl")
            .arg("get")
            .arg(&qualified_names_joined)
            .arg("--all-namespaces")
            .arg("-o")
            .arg("yaml")
    )
}

fn map_to_resource_type(map: &HashMap<String, String>) -> ApiResourceType {
    ApiResourceType {
        name: String::from(&map["NAME"]),
        short_names: util::string::split_to_vec(&map["SHORTNAMES"], ",", true),
        namespaced: bool::from_str(&map["NAMESPACED"]).expect("non-bool 'NAMESPACED' in map"),
        kind: String::from(&map["KIND"]),
        verbs: util::string::split_to_vec_r(util::string::remove_wrap(&map["VERBS"]), &util::regex::Regex::new(r"\s+").expect("BUG: invalid regex to split VERBS"), true),
        api_version: super::parse_api_version(&map["APIVERSION"]),
    }
}

File: internal/k8s/kubectl.go
package k8s

import (
	"context"
	"fmt"
	"github.com/GiGurra/cmder"
	"github.com/GiGurra/kdump/internal/errh"
	"github.com/samber/lo"
	"io"
	"log/slog"
	"os/exec"
	"strings"
	"time"
)

func commandExists(command string) bool {
	_, err := exec.LookPath(command)
	return err == nil
}

func CheckNecessaryCliAppsAvailable() {

	// Check we have
	slog.Info("Checking that kubectl is installed...")
	if !commandExists("kubectl") {
		panic("kubectl not on path!")
	}

	slog.Info("Checking that kubectl neat is installed...")
	RunCommand("kubectl", "neat", "--help")
}

func ListAvailableContexts() []string {
	allString := RunCommand("kubectl", "config", "get-contexts", "-o", "name")
	lines := strings.Split(allString, "\n")
	trimmedLines := lo.Map(lines, func(in string, _ int) string { return strings.TrimSpace(in) })
	return lo.Filter(trimmedLines, func(in string, _ int) bool { return len(in) > 0 })
}

func CurrentContext() string {
	return strings.TrimSpace(RunCommand("kubectl", "config", "current-context"))
}

func DownloadEverything(types []*ApiResourceType) string {

	qualifiedTypeNames := lo.Map(types, func(in *ApiResourceType, _ int) string {
		return in.QualifiedName
	})

	return RunCommand("kubectl", "get", strings.Join(qualifiedTypeNames, ","), "--all-namespaces", "-o", "yaml")
}

type ApiResourceTypesResponse struct {
	All        []*ApiResourceType
	Accessible ApiResourceTypesAccessible
}

type ApiResourceTypesAccessible struct {
	All        []*ApiResourceType
	Global     []*ApiResourceType
	Namespaced []*ApiResourceType
}

func ApiResourceTypes() ApiResourceTypesResponse {

	slog.Info("Checking what api resource types are available...")

	rawString := RunCommand("kubectl", "api-resources", "-o", "wide")

	apiResourcesRaw := ParseStdOutTable(rawString)

	allApiResources := lo.Map(apiResourcesRaw, func(in map[string]string, _ int) *ApiResourceType {

		out := &ApiResourceType{
			Name:       MapStrValOrElse(in, "NAME", ""),
			ShortNames: CsvStr2arr(MapStrValOrElse(in, "SHORTNAMES", "")),
			Namespaced: Str2boolOrElse(MapStrValOrElse(in, "NAMESPACED", ""), false),
			Kind:       MapStrValOrElse(in, "KIND", ""),
			Verbs:      WierdKubectlArray2arr(MapStrValOrElse(in, "VERBS", "")),
		}

		apiVersionString := MapStrValOrElse(in, "APIVERSION", "")
		parts := strings.Split(apiVersionString, "/")
		if len(parts) == 1 {
			out.ApiVersion = ApiVersion{Version: parts[0]}
			out.QualifiedName = out.Name
		} else if len(parts) == 2 {
			out.ApiVersion = ApiVersion{Version: parts[1], Name: parts[0]}
			out.QualifiedName = out.Name + "." + out.ApiVersion.Name
		} else {
			panic("Unable to parse ApiVersion from string: " + apiVersionString)
		}

		return out

	})

	accessibleApiResources := lo.Filter(allApiResources, func(r *ApiResourceType, _ int) bool { return lo.Contains(r.Verbs, "get") })
	globalResources := lo.Filter(accessibleApiResources, func(r *ApiResourceType, _ int) bool { return !r.Namespaced })
	namespacedResources := lo.Filter(accessibleApiResources, func(r *ApiResourceType, _ int) bool { return r.Namespaced })

	return ApiResourceTypesResponse{
		All: allApiResources,
		Accessible: ApiResourceTypesAccessible{
			All:        accessibleApiResources,
			Global:     globalResources,
			Namespaced: namespacedResources,
		},
	}
}

func RunCommand(app string, args ...string) string {

	fullCommand := app + " " + strings.Join(args, " ")

	res := cmder.
		NewA(app, args...).
		WithTotalTimeout(30 * time.Minute).
		Run(context.Background())

	if res.Err != nil {
		panic(fmt.Sprintf(`command "%s" failed with error code %d: %s`, fullCommand, res.ExitCode, res.Combined))
	}

	return strings.TrimSpace(res.StdOut)
}

func PipeToCommand(input string, app string, args ...string) string {

	subProcess := exec.Command(app, args...)
	stdin := errh.Unwrap(subProcess.StdinPipe()) // stdin must be opened before .start/.output

	go func() {
		errh.Unwrap(io.WriteString(stdin, input))
		errh.Ignore(stdin.Close())
	}()

	outputBytes := errh.Unwrap(subProcess.Output()) // starts and awaits the result of the child process

	return strings.TrimSpace(string(outputBytes))
}
